src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/MissingWorkloadRegistrationService.cs

```diff
@@ -0,0 +1,134 @@
+    /// </summary>
+    [Export(typeof(IMissingWorkloadRegistrationService))]
+    internal class MissingWorkloadRegistrationService : OnceInitializedOnceDisposedAsync, IMissingWorkloadRegistrationService
+    {
+        private readonly IDictionary<Guid, ISet<WorkloadDescriptor>> _projectGuidToWorkloadDescriptorsMap;
+        private readonly IDictionary<Guid, ISet<ProjectConfiguration>> _projectGuidToProjectConfigurationsMap;
+        private readonly IVsService<SVsBrokeredServiceContainer, IBrokeredServiceContainer> _serviceBrokerContainer;
+        private readonly IVisualStudioComponentIdTransformer _visualStudioComponentIdTransformer;
+
+        [ImportingConstructor]
+        public MissingWorkloadRegistrationService(
+            JoinableTaskContext joinableTaskContext,
+            IVsService<SVsBrokeredServiceContainer, IBrokeredServiceContainer> serviceBrokerContainer,
+            IProjectThreadingService projectThreadingService,
+            IVisualStudioComponentIdTransformer visualStudioComponentIdTransformer)
+            : base(new JoinableTaskContextNode(joinableTaskContext))
+        {
+            _projectGuidToWorkloadDescriptorsMap = new Dictionary<Guid, ISet<WorkloadDescriptor>>();
+            _projectGuidToProjectConfigurationsMap = new Dictionary<Guid, ISet<ProjectConfiguration>>();
+            _serviceBrokerContainer = serviceBrokerContainer;
+            _visualStudioComponentIdTransformer = visualStudioComponentIdTransformer;
+        }
+
+        public void ClearMissingWorkloadMetadata()
+        {
+            _projectGuidToWorkloadDescriptorsMap.Clear();
+        }
+
+        public Task RegisterMissingWorkloadAsync(Guid projectGuid, ProjectConfiguration projectConfiguration, WorkloadDescriptor workloadDescriptor, CancellationToken cancellationToken)
+        {
+            if (!WorkloadDescriptor.Empty.Equals(workloadDescriptor))
+            {
+                if (!_projectGuidToWorkloadDescriptorsMap.TryGetValue(projectGuid, out var workloadDescriptorSet))
+                {
+                    workloadDescriptorSet = new HashSet<WorkloadDescriptor>();
+                    _projectGuidToWorkloadDescriptorsMap[projectGuid] = workloadDescriptorSet;
+                }
+
+                workloadDescriptorSet.Add(workloadDescriptor);
+            }
+
+            if (_projectGuidToProjectConfigurationsMap.TryGetValue(projectGuid, out var projectConfigurationSet))
+            {
+                projectConfigurationSet.Remove(projectConfiguration);
+                if (projectConfigurationSet.Count == 0 && ShouldDisplayMissingComponentsPrompt())
+                {
+                    return DisplayMissingComponentsPromptAsync(cancellationToken);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void RegisterProjectConfiguration(Guid projectGuid, ProjectConfiguration projectConfiguration)
```

Comment:
I don't see a way to de-register these.

What happens if a configuration is removed? For example, if I change a target framework in a multi-targeting project.

What happens if I remove a project from the solution?

Reply:
> I don't see a way to de-register these.

I have added an explicit Unregister method to the service for the MissingWorkloadDetector. Thanks for the feedback.